mixnet/node/src/lib.rs

```diff
@@ -77,205 +76,279 @@ impl MixnetNode {
-        }
+
+        self.message_tx.send(TargetedMessage::new(addr, msg))?;
         Ok(())
     }
 }
+
+struct MessageHandler {
+    // TODO: remove this allow when we implement the retry logic
+    #[allow(dead_code)]
+    config: MixnetNodeConfig,
+    message_rx: mpsc::UnboundedReceiver<TargetedMessage>,
+    message_tx: mpsc::UnboundedSender<TargetedMessage>,
+    connections: HashMap<SocketAddr, TcpStream>,
+}
+
+impl MessageHandler {
+    pub fn new(
+        message_tx: mpsc::UnboundedSender<TargetedMessage>,
+        message_rx: mpsc::UnboundedReceiver<TargetedMessage>,
+        config: MixnetNodeConfig,
+    ) -> Self {
+        Self {
+            message_tx,
+            message_rx,
+            connections: HashMap::with_capacity(config.connection_pool_size),
+            config,
+        }
+    }
+
+    pub async fn run(mut self) {
+        loop {
+            tokio::select! {
+                msg = self.message_rx.recv() => {
+                    if let Some(msg) = msg {
+                        match self.send(msg).await {
+                            Ok(Some(msg)) => {
+                                let _ = self.message_tx.send(msg);
+                            }
+                            Ok(None) => {},
+                            Err(e) => {
+                                tracing::error!("failed to send msg: {e}");
+                            },
+                        }
+                    } else {
+                        // Channel closed, we should shutdown the message handler thread
+                        return;
+                    }
+                },
+                _ = tokio::signal::ctrl_c() => {
+                    tracing::info!("Shutting down message handler thread...");
+                    return;
+                }
+            }
+        }
+    }
+
+    /// Send a message to the remote node,
+    /// return Ok(Some((Duration, Message))) if the message is not sent and the error is retryable
+    /// return Ok(None) if the message is sent successfully
+    /// return Err(e) if the message is not sent and the error is not retryable
+    async fn send(&mut self, mut msg: TargetedMessage) -> Result<Option<TargetedMessage>> {
+        if msg.retry_count >= self.config.max_retries {
+            return Err(MixnetNodeError::Protocol(ProtocolError::ReachMaxRetries(
+                self.config.max_retries,
+            )));
+        }
+        use std::io::ErrorKind;
+
+        if let std::collections::hash_map::Entry::Vacant(e) = self.connections.entry(msg.target) {
+            match TcpStream::connect(msg.target).await {
+                Ok(tcp) => {
+                    e.insert(tcp);
+                }
+                Err(e) => {
+                    tracing::error!("failed to connect to {}: {e}", msg.target);
+                    return Ok(Some(msg));
+                }
+            }
+        }
+
+        let tcp = self.connections.get_mut(&msg.target).unwrap();
```

Comment:
`let tcp = self.connections.entry(msg.target).or_insert_with(|| {TcpStream::connect(...)}).`

Reply:
This seems cannot be done by this way, `TcpStream::connect` is an async fn.